ec.1
.\" Man page generated from reStructuredText.
.
.TH EC 1 "2016-09-30" "1.2.2" ""
.SH NAME
ec \- engineering calculator
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" :Copyright: public domain
.
.\" :Manual group: Utilities
.
.SH SYNOPSIS
.sp
\fBec\fP [\fIoptions\fP] [\fIscripts\fP ...]
.SH OPTIONS
.INDENT 0.0
.TP
.B \-i\fP,\fB  \-\-interactive
Open an interactive session.
.TP
.BI \-s \ <file>\fP,\fB \ \-\-startup \ <file>
Run commands from file to initialize calculator before
any script or interactive session is run, stack is
cleared after it is run.
.TP
.B \-c\fP,\fB  \-\-nocolor
Do not use colors in the output.
.TP
.B \-v\fP,\fB  \-\-verbose
Narrate the execution of any scripts.
.TP
.B \-V\fP,\fB  \-\-version
Print the ec version information.
.TP
.B \-h\fP,\fB  \-\-help
Print the usage and exit.
.UNINDENT
.SH DESCRIPTION
.sp
\fBec\fP is a stack\-based (RPN) engineering calculator with a text\-based user
interface that is intended to be used interactively.
.sp
If run with no arguments, an interactive session is started.  If arguments
are present, they are tested to see if they are filenames, and if so, the
files are opened and the contents are executed as a script.  If they are not
file names, then the arguments themselves are treated as scripts and
executed directly. The scripts are run in the order they are specified.  In
this case an interactive session would not normally be started, but if the
interactive option is specified, it would be started after all scripts have
been run.
.sp
The contents of \fI~/.ecrc\fP, \fI\&./.ecrc\fP, and the start up file will be run upon
start up if they exist, and then the stack is cleared.
.SH STACK
.sp
As you enter numbers they are pushed onto a stack.  The most recent member
of the stack is referred to as the \fIx\fP register and the second most recent
is the \fIy\fP register.  All other members of the stack are unnamed.  Operators
consume numbers off the stack to use as operands and then they push the
results back on the stack.  The operations are performed immediately and
there is no use of parentheses to group calculations.  Any intermediate
results are stored on the stack until needed.  For example,
.INDENT 0.0
.INDENT 3.5
.nf
4
6
+
.fi
.sp
.UNINDENT
.UNINDENT
.sp
In this case 4 gets pushed on to the stack first to become \fIx\fP\&. Then 6 gets
pushed on to the stack to become \fIx\fP, which makes 4 \fIy\fP\&. Finally, + pulls
both off the stack, sums them, and then pushes the result of 10 back onto
the stack.  The stack is left with only one number on it, 10.
.sp
After each line \fBec\fP responds by printing the value of the \fIx\fP register.
Thus the above example would actually look like this:
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 4
\fB4\fP: 6
\fB6\fP: +
\fB10\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
The benefit of the stack is that it allows you to easily store temporary
results while you perform your calculation. For example, to evaluate (34 \-
61)*(23 \- 56) you would use:
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 34
\fB34\fP: 61
\fB61\fP: \-
\fB\-27\fP: 23
\fB23\fP: 56
\fB56\fP: \-
\fB\-33\fP: *
\fB891\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
Notice that you entered the numbers as you saw them in the formula you were
evaluating, and there was no need to enter parentheses, however the
operators were rearranged in order to express the precedence of the
operations.
.sp
It is not necessary to type enter after each number or operator. You can
combine them onto one line and just type enter when you would like to see
the result:
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 34 61 \- 23 56 \- *
\fB891\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
Furthermore, it is not necessary to type a space between a number and most
operators. For example, the above could be entered as:
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 34 61\- 23 56\- *
\fB891\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
You can print the entire stack using \fIstack\fP, and clear it using \fIclstack\fP\&.
For example,
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 1 2 3 stack
.in +2
.in +2
 3
.in -2
\fIy\fP: 2
\fIx\fP: 1
.in -2
\fB3\fP: clstack
\fB0\fP: stack
\fB0\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.SH REAL NUMBERS
.sp
Numbers can be entered using normal integer, floating point, and scientific
notations. For example,
.INDENT 0.0
.INDENT 3.5
.nf
42
3.141592
5,439,749.97
2.998e8
13.80651e\-24
.fi
.sp
.UNINDENT
.UNINDENT
.sp
In addition, you can also use the normal SI scale factors to represent
either large or small numbers without using scientific notation.
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B Y
1e24 (yotta)
.TP
.B Z
1e21 (zetta)
.TP
.B E
1e18 (exa)
.TP
.B P
1e15 (peta)
.TP
.B T
1e12 (terra)
.TP
.B G
1e9 (giga)
.TP
.B M
1e6 (mega)
.TP
.B k, K
1e3 (kilo)
.TP
.B _
unity (1)
.TP
.B m
1e\-3 (milli)
.TP
.B u
1e\-6 (micro)
.TP
.B n
1e\-9 (nano)
.TP
.B p
1e\-12 (pico)
.TP
.B f
1e\-15 (fempto)
.TP
.B a
1e\-18 (atto)
.TP
.B z
1e\-21 (zepto)
.TP
.B y
1e\-24 (yocto)
.UNINDENT
.UNINDENT
.UNINDENT
.sp
For example, 10M represents 1e7 and 8.8p represents 8.8e\-12.
.sp
Optionally, numbers can be combined with simple units. For example,
.INDENT 0.0
.INDENT 3.5
.nf
10KHz
3.16pF
2.5_V
4.7e\-10F
.fi
.sp
.UNINDENT
.UNINDENT
.sp
Units are only allowed after a scale factor or an exponent, and the units
are optional. In this way, 1m represents 1e\-3 rather than 1 meter. If you
wish to enter 1 meter, use 1_m. The underscore is the unity scale factor.
.sp
In this case the units must be simple identifiers (must not contain special
characters). For complex units, such as "rads/s", or for numbers that do not
have scale factors, it is possible to attach units to a number in the \fIx\fP
register by entering a quoted string.
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 6.626e\-34
\fB662.6e\-36\fP: "J\-s"
\fB662.6e\-36 J\-s\fP: 50k "V/V"
\fB50 KV/V\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
The dollar sign ($) is a special unit that is given before the number.
.INDENT 0.0
.INDENT 3.5
.nf
$100K
.fi
.sp
.UNINDENT
.UNINDENT
.sp
\fBec\fP takes a conservative approach to units. You can enter them and it
remembers them, but they do not survive any operation where the resulting
units would be in doubt.  In this way it displays units when it can, but
should never display incorrect or misleading units. For example:
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 100MHz
\fB100 MHz\fP: 2pi*
\fB628.32M\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
You can display real numbers using one of three available formats, \fIfix\fP,
\fIsci\fP, or \fIeng\fP\&. These display numbers using fixed point notation (a fixed
number of digits to the right of the decimal point), scientific notation (a
mantissa and an exponent), and engineering notation (a mantissa and an SI
scale factor).  You can optionally give an integer immediately after the
display mode to indicate the desired precision.  For example,
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 1000
\fB1K\fP: fix2
\fB1000.00\fP: sci3
\fB1.000e+03\fP: eng4
\fB1K\fP: 2pi*
\fB6.2832K\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
Notice that scientific notation always displays the specified number of
digits whereas engineering notation suppresses zeros at the end of the
number.
.sp
When displaying numbers using engineering notation, \fBec\fP does not use the
full range of available scale factors under the assumption that the largest
and smallest would be unfamiliar to most people. For this reason, \fBec\fP
only uses the most common scale factors when outputting numbers (T, G, M, K,
m, u, n, p, f, a).
.SH INTEGERS
.sp
You can enter integers in either hexadecimal (base 16), decimal (base 10),
octal (base 8), or binary (base 2). You can use either programmers notation
(leading 0) or Verilog notation (leading \(aq) as shown in the examples below:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B 0xFF
hexadecimal
.TP
.B 99
decimal
.TP
.B 0o77
octal
.TP
.B 0b1101
binary
.TP
.B \(aqhFF
Verilog hexadecimal
.TP
.B \(aqd99
Verilog decimal
.TP
.B \(aqo77
Verilog octal
.TP
.B \(aqb1101
Verilog binary
.UNINDENT
.UNINDENT
.UNINDENT
.sp
Internally, \fBec\fP represents all numbers as double\-precision real numbers.
To display them as decimal integers, use \fIfix0\fP\&. However, you can display
the numbers in either base 16 (hexadecimal), base 10 (decimal), base 8
(octal) or base 2 (binary)  by setting the display mode.  Use either \fIhex\fP,
\fIfix0\fP, \fIoct\fP, \fIbin\fP, \fIvhex\fP, \fIvdec\fP, \fIvoct\fP, or \fIvbin\fP\&. In each of
these cases the number is rounded to the closest integer before it is
displayed. Add an integer after the display mode to control the number of
digits. For example:
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 1000
\fB1K\fP: hex
\fB0x3b8\fP: hex8
\fB0x000003b8\fP: hex0
\fB0x3b8\fP: voct
\fB\(aqo1750\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.SH COMPLEX NUMBERS
.sp
\fBec\fP provides limited support for complex numbers. Two imaginary constants
are available that can be used to construct complex numbers, \fIj\fP and
\fIj2pi\fP\&. In addition, two functions are available for converting complex
numbers to real, \fImag\fP returns the magnitude and \fIph\fP returns the phase.
They are unusual in that they do not replace the value in the \fIx\fP register
with the result, instead they simply push either the magnitude of phase into
the \fIx\fP register, which pushes the original complex number into the \fIy\fP
register. For example,
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 1 j +
\fB1 + j\fP: mag
\fB1.4142\fP: pop
\fB1 + j\fP: ph
\fB45 degs\fP: stack
.in +2
\fIy\fP: 1 + j
\fIx\fP: 45 degs
.in -2
\fB45 degs\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
You can also add the imaginary unit to real number constants. For example,
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: j10M
\fBj10M\fP: \-j1u *
\fB10\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
Only a small number of functions actually support complex numbers; currently
only \fIexp\fP and \fIsqrt\fP\&. However, most of the basic arithmetic operators
support complex numbers.
.SH CONSTANTS
.sp
\fBec\fP provides several useful mathematical and physical constants that are
accessed by specifying them by name. The physical constants are given in MKS
units. The available constants include:
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B pi
the ratio of a circle\(aqs circumference to its diameter (rads)
.TP
.B 2pi
the ratio of a circle\(aqs circumference to its radius (rads)
.TP
.B rt2
square root of two
.TP
.B 0C
0 Celsius in Kelvin (K)
.TP
.B j
imaginary unit (square root of \-1)
.TP
.B j2pi
j*2*pi (rads)
.TP
.B k
Boltzmann constant (J/K)
.TP
.B h
Planck constant (J\-s)
.TP
.B q
elementary charge (the charge of an electron) (C)
.TP
.B c
speed of light in a vacuum (m/s)
.TP
.B eps0
permittivity of free space (F/m)
.TP
.B mu0
permeability of free space (H/m)
.TP
.B Z0
Characteristic impedance of free space (Ohms)
.TP
.B hbar
Reduced Planck constant (J\-s)
.TP
.B me
mass of an electron (g)
.TP
.B mp
mass of a proton (g)
.TP
.B G
universal gravitational constant (m^3/(kg\-s^2))
.TP
.B g
standard acceleration of gravity (m/s^2)
.TP
.B R
molar gas constant (J/(mol\-K))
.TP
.B NA
Avagadro Number (/mol)
.TP
.B rand
random number between 0 and 1
.UNINDENT
.UNINDENT
.UNINDENT
.sp
As an example of using the predefined constants, consider computing the
thermal voltage, kT/q.
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: k 27 0C + * q/
\fB25.865m\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.SH VARIABLES
.sp
You can store the contents of the \fIx\fP register to a variable by using an
equal sign followed immediately by the name of the variable. To recall it,
simply use the name. For example,
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 100MHz =freq
\fB100 MHz\fP: 2pi* "rads/s" =omega
\fB628.32 Mrads/s\fP: 1pF =cin
\fB1 pF\fP: 1 omega cin* /
\fB1.5915K\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
You can display all known variables using \fIvars\fP\&. If you did so immediately
after entering the lines above, you would see:
.INDENT 0.0
.INDENT 3.5
.nf
\fB1.5915K\fP: vars
.in +2
\fIRref\fP: 50 Ohms
\fIcin\fP: 1 pF
\fIfreq\fP: 100 MHz
\fIomega\fP: 628.32 Mrads/s
.in -2
.fi
.sp
.UNINDENT
.UNINDENT
.sp
Choosing a variable name that is the same as a one of a built\-in command or
constant causes the built\-in name to be overridden. Be careful when doing
this as once a built\-in name is overridden it can no longer be accessed.
.sp
Notice that a variable \fIRref\fP exists that you did not create. This is a
predefined variable that is used in dBm calculations. You are free to change
its value if you like.
.SH USER-DEFINED FUNCTIONS
.sp
You can define functions in the following way:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
( ... )name
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Here \(aq(\(aq starts the function definition and \(aq)name\(aq ends it. The name must
be immediately adjacent to the name. The \(aq...\(aq represents a sequence of
calculator actions. For example:
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: (2pi * "rads/s")to_omega
\fB0\fP: (2pi / "Hz")to_freq
\fB0\fP: 100MHz
\fB100 MHz\fP: to_omega
\fB628.32 Mrads/s\fP: to_freq
\fB100 MHz\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
The actions entered while defining the function are not evaluated until the
function itself is evaluated.
.sp
Once defined, you can review your function with the \fIvars\fP command. It shows
both the variable and the function definitions:
.INDENT 0.0
.INDENT 3.5
.nf
\fIRref\fP: 50 Ohms
\fIto_freq\fP: (2pi / "Hz")
\fIto_omega\fP: (2pi * "rads/s")
.fi
.sp
.UNINDENT
.UNINDENT
.sp
The value of the functions are delimited with parentheses.
.SH OPERATORS, FUNCTIONS, NUMBERS AND COMMANDS
.sp
In the following descriptions, optional values are given in brackets ([])
and values given in angle brackets (<>) are not to be taken literally (you
are expected to choose a suitable value). For example "fix[<\fIN\fP>]" can
represent "fix" or "fix4", but not "fixN".
.sp
For each action that changes the stack a synopsis of those changes is given
in the form of two lists separated by \fB=>\fP\&. The list on the left
represents the stack before the action is applied, and the list on the right
represents the stack after the action was applied. In both of these lists,
the \fIx\fP register is given first (on the left). Those registers that are
involved in the action are listed explicitly, and the rest are represented
by \fB\&...\fP\&. In the before picture, the names of the registers involved in
the action are simply named. In the after picture, the new values of the
registers are described. Those values represented by \fB\&...\fP on the right
side of \fB=>\fP are the same as represented by \fB\&...\fP on the left, though
they may have moved. For example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, y, ... => x+y, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This represents addition. In this case the values in the \fIx\fP and \fIy\fP
registers are summed and placed into the \fIx\fP register. All other values move
to the left one place.
.SS Arithmetic Operators
.sp
\fB+\fP: addition
.INDENT 0.0
.INDENT 3.5
The values in the \fIx\fP and \fIy\fP registers are popped from the
stack and the sum is placed back on the stack into the \fIx\fP
register.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, y, ... => x+y, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fB\-\fP: subtraction
.INDENT 0.0
.INDENT 3.5
The values in the \fIx\fP and \fIy\fP registers are popped from the
stack and the difference is placed back on the stack into the \fIx\fP
register.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, y, ... => x\-y, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fB*\fP: multiplication
.INDENT 0.0
.INDENT 3.5
The values in the \fIx\fP and \fIy\fP registers are popped from the
stack and the product is placed back on the stack into the \fIx\fP
register.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, y, ... => x*y, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fB/\fP: true division
.INDENT 0.0
.INDENT 3.5
The values in the \fIx\fP and \fIy\fP registers are popped from the stack and
the quotient is placed back on the stack into the \fIx\fP register.  Both
values are treated as real numbers and the result is a real number. So
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 1 2/
\fB500m\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, y, ... => y/x, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fB//\fP: floor division
.INDENT 0.0
.INDENT 3.5
The values in the \fIx\fP and \fIy\fP registers are popped from the
stack, the quotient is computed and then converted to an integer using
the floor operation (it is replaced by the largest integer that is
smaller than the quotient), and that is placed back on the stack into
the \fIx\fP register.  So
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 1 2//
\fB0\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, y, ... => y//x, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fB%\fP: modulus
.INDENT 0.0
.INDENT 3.5
The values in the \fIx\fP and \fIy\fP registers are popped from the stack, the
quotient is computed and the remainder is placed back on the stack into
the \fIx\fP register.  So
.INDENT 0.0
.INDENT 3.5
.nf
\fB0\fP: 14 3%
\fB2\fP:
.fi
.sp
.UNINDENT
.UNINDENT
.sp
In this case 2 is the remainder because 3 goes evenly into 14 three
times, which leaves a remainder of 2.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, y, ... => y%x, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBchs\fP: change sign
.INDENT 0.0
.INDENT 3.5
The value in the \fIx\fP register is replaced with its negative.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, ... => \-x, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBrecip\fP: reciprocal
.INDENT 0.0
.INDENT 3.5
The value in the \fIx\fP register is replaced with its reciprocal.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, ... => 1/x, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBceil\fP: round towards positive infinity
.INDENT 0.0
.INDENT 3.5
The value in the \fIx\fP register is replaced with its value rounded
towards infinity (replaced with the smallest integer greater than its
value).
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, ... => ceil(x), ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBfloor\fP: round towards negative infinity
.INDENT 0.0
.INDENT 3.5
The value in the \fIx\fP register is replaced with its value rounded
towards negative infinity (replaced with the largest integer smaller
than its value).
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, ... => floor(x), ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fB!\fP: factorial
.INDENT 0.0
.INDENT 3.5
The value in the \fIx\fP register is replaced with its factorial.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, ... => x!, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fB%chg\fP: percent change
.INDENT 0.0
.INDENT 3.5
The values in the \fIx\fP and \fIy\fP registers are popped from the stack and
the percent difference between \fIx\fP and \fIy\fP relative to \fIy\fP is pushed
back into the \fIx\fP register.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, y, ... => 100*(x\-y)/y, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fB||\fP: parallel combination
.INDENT 0.0
.INDENT 3.5
The values in the \fIx\fP and \fIy\fP registers are popped from the stack and
replaced with the reciprocal of the sum of their reciprocals.  If the
values in the \fIx\fP and \fIy\fP registers are both resistances, both
elastances, or both inductances, then the result is the resistance,
elastance or inductance of the two in parallel. If the values are
conductances, capacitances or susceptances, then the result is the
conductance, capacitance or susceptance of the two in series.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, y, ... => 1/(1/x+1/y), ...
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.SS Powers, Roots, Exponentials and Logarithms
.sp
\fB**\fP: raise y to the power of x
.INDENT 0.0
.INDENT 3.5
The values in the \fIx\fP and \fIy\fP registers are popped from the
stack and replaced with the value of \fIy\fP raised to the power of
\fIx\fP\&.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, y, ... => y**x, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
aliases: pow, ytox
.UNINDENT
.UNINDENT
.sp
\fBexp\fP: natural exponential
.INDENT 0.0
.INDENT 3.5
The value in the \fIx\fP register is replaced with its exponential.
Supports a complex argument.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, ... => exp(x), ...
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
alias: powe
.UNINDENT
.UNINDENT
.sp
\fBln\fP: natural logarithm
.INDENT 0.0
.INDENT 3.5
The value in the \fIx\fP register is replaced with its natural logarithm.
Supports a complex argument.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, ... => ln(x), ...
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
alias: loge
.UNINDENT
.UNINDENT
.sp
\fBpow10\fP: raise 10 to the power of x
.INDENT 0.0
.INDENT 3.5
The value in the \fIx\fP register is replaced with 10 raised to \fIx\fP\&.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, ... => 10**x, ...
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
alias: 10tox
.UNINDENT
.UNINDENT
.sp
\fBlog\fP: base 10 logarithm
.INDENT 0.0
.INDENT 3.5
The value in the \fIx\fP register is replaced with its common logarithm.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
x, ... => log(x), ...
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
aliases: log10, lg
.UNINDENT
.UNINDENT
.sp
\fBpow2\fP: raise 2 to the power of x
.INDENT 0.0
.INDENT 3.5

ec.py
#!/usr/bin/env python
#
# Engineering Calculator {{{1
#
# An RPN calculator that supports numbers with SI scale factors and units.
#
# Copyright (C) 2013 Kenneth S. Kundert
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see [http://www.gnu.org/licenses/].

# Imports {{{1
from __future__ import division
from actions import (
    actionsToUse, predefinedVariables, defaultFormat, defaultDigits
)
from calculator import Calculator, Display, CalculatorError
from cmdline import commandLineProcessor
from os.path import expanduser
import sys, os

# Configure the command line processor {{{1
clp = commandLineProcessor()
clp.setDescription('Engineering Calculator', '\n'.join([
    'A stack-based (RPN) engineering calculator with a text-based user '
  , 'interface that is intended to be used interactively.'
  , ''
  , 'If run with no arguments, an interactive session is started. '
  , 'If arguments are present, they are tested to see if they are '
  , 'filenames, and if so, the files are opened and the contents are '
  , 'executed as a script.  If they are not file names, then the arguments '
  , 'themselves are treated as scripts and executed directly. The scripts '
  , 'are run in the order they are specified.  In this case an interactive '
  , 'session would not normally be started, but if the interactive option '
  , 'is specified, it would be started after all scripts have been run.'
  , ''
  , 'The contents of ~/.ecrc, ./.ecrc, and the start up file will be run '
  , 'upon start up if they exist, and then the stack is cleared.'
]))
clp.setNumArgs((0,), '[scripts ...]')
clp.setHelpParams(key='--help', colWidth=18)
opt = clp.addOption(key='interactive', shortName='i', longName='interactive')
opt.setSummary('Open an interactive session.')
opt = clp.addOption(key='printx', shortName='x', longName='printx')
opt.setSummary(' '.join([
    'Print value of x register upon termination,'
  , 'ignored with interactive sessions.'
]))
opt = clp.addOption(key='startup', shortName='s', longName='startup')
opt.setSummary(' '.join([
    'Run commands from file to initialize calculator before any script or'
  , 'interactive session is run, stack is cleared after it is run.'
]))
opt.setNumArgs(1, 'file')
opt = clp.addOption(key='nocolor', shortName='c', longName='nocolor')
opt.setSummary('Do not color the output.')
opt = clp.addOption(key='verbose', shortName='v', longName='verbose')
opt.setSummary('Narrate the execution of any scripts.')
opt = clp.addOption(
    key='help', shortName='h', longName='help', action=clp.printHelp
)
opt.setSummary('Print usage information.')

# Process the command line {{{1
clp.process()

# get the command line options and arguments
opts = clp.getOptions()
args = clp.getArguments()
progName = clp.progName()
colorize = 'nocolor' not in opts
startUpFile = opts.get('startup', [])
interactiveSession = 'interactive' in opts or not args
printXuponTermination = 'printx' in opts
verbose = 'verbose' in opts

# Import and configure the text colorizer {{{1
if colorize:
    try:
        import textcolors
    except ImportError:
        colorize = False

if colorize:
    colors = textcolors.Colors()
    error = colors.colorizer('red')
    highlight = colors.colorizer('magenta')
    warning = colors.colorizer('yellow')
else:
    error = highlight = warning = lambda x: x

# Define utility functions {{{1
def printWarning(message):
    print("%s: %s" % (warning('Warning'), message))

def evaluateLine(calc, line, prompt):
    try:
        result = calc.evaluate(
            calc.split(line)
        )
        prompt = calc.format(result)
    except CalculatorError as err:
        if interactiveSession:
            print(error(err.message))
            prompt = calc.restoreStack()
        else:
            sys.exit(error(err.message))
    return prompt

# Create calculator {{{1
calc = Calculator(
    actionsToUse
  , Display(format=defaultFormat, digits=defaultDigits)
  , predefinedVariables
  , backUpStack=interactiveSession
  , warningPrinter=printWarning
)
prompt = '0'

# Run start up files {{{1
rcFiles = ['%s/.ecrc' % each for each in ['~', '.']]
for each in rcFiles + startUpFile:
    lineno = None
    try:
        cmdFile = expanduser(each)
        with open(cmdFile) as pFile:
            for lineno, line in enumerate(pFile):
                prompt = evaluateLine(calc, line, prompt)
                if verbose:
                    print("%s %s: %s ==> %s" % (
                        cmdFile, lineno, line.strip(), prompt
                    ))
    except IOError as err:
        if err.errno != 2 or each not in rcFiles:
            sys.exit('%s%s: %s: %s' % (
                each
              , ('.%s' % lineno+1) if lineno != None else ''
              , err.filename
              , err.strerror
            ))
calc.stack.clear()
prompt = '0'

# Run scripts {{{1
for arg in args:
    try:
        cmdFile = expanduser(arg)
        if os.path.exists(cmdFile):
            with open(cmdFile) as pFile:
                for lineno, line in enumerate(pFile):
                    loc = '%s.%s: ' % (cmdFile, lineno+1)
                    prompt = evaluateLine(calc, line, prompt)
                    if verbose:
                        print("%s %s: %s ==> %s" % (
                            cmdFile, lineno, line.strip(), prompt
                        ))
        else:
            loc = ''
            prompt = evaluateLine(calc, arg, prompt)
            if verbose:
                print("%s ==> %s" % (line, prompt))
    except IOError as err:
        if err.errno != 2:
            sys.exit('%s: %s' % (err.filename, err.strerror))

# Interact with user {{{1
if (interactiveSession):
    while(True):
        try:
            entered = raw_input('%s: ' % highlight(prompt)) # python 2
        except (EOFError, KeyboardInterrupt):
            print()
            sys.exit(0)
        except NameError:
            try:
                entered = input('%s: ' % highlight(prompt)) # python 3
            except (EOFError, KeyboardInterrupt):
                print()
                sys.exit(0)
        prompt = evaluateLine(calc, entered, prompt)
elif printXuponTermination:
    print(prompt)
sys.exit(0)
